vmdk_plugin/drivers/vmdk/vmdk_driver.go

```diff
@@ -349,75 +425,43 @@ func (d *VolumeDriver) Path(r volume.Request) volume.Response {
-	// since the volume will have been never mounted.
-	// Note: for new keys, GO maps return zero value, so no need for if_exists.
-
-	refcnt := d.incrRefCount(r.Name) // save map traversal
-	log.Debugf("volume name=%s refcnt=%d", r.Name, refcnt)
-	if refcnt > 1 {
-		log.WithFields(
-			log.Fields{"name": r.Name, "refcount": refcnt},
-		).Info("Already mounted, skipping mount. ")
-		return volume.Response{Mountpoint: getMountPoint(r.Name)}
-	}
-
-	// This is the first time we are asked to mount the volume, so comply
-	status, err := d.ops.Get(r.Name)
-
-	fstype := fs.FstypeDefault
-	isReadOnly := false
-	if err != nil {
-		d.decrRefCount(r.Name)
-		return volume.Response{Err: err.Error()}
-	}
-	// Check access type.
-	value, exists := status["access"].(string)
-	if !exists {
-		msg := fmt.Sprintf("Invalid access type for %s, assuming read-write access.", r.Name)
-		log.WithFields(log.Fields{"name": r.Name, "error": msg}).Error("")
-		isReadOnly = false
-	} else if value == "read-only" {
-		isReadOnly = true
-	}
-
-	// Check file system type.
-	value, exists = status["fstype"].(string)
-	if !exists {
-		msg := fmt.Sprintf("Invalid filesystem type for %s, assuming type as %s.",
-			r.Name, fstype)
-		log.WithFields(log.Fields{"name": r.Name, "error": msg}).Error("")
-		// Fail back to a default version that we can try with.
-		value = fs.FstypeDefault
+	// lock the state
+	d.refCounts.StateMtx.Lock()
+	defer d.refCounts.StateMtx.Unlock()
+	if d.refCounts.GetSuccess() != true {
+		// complete the mount process under the lock to avoid races
+		// dont do refcounting. Just mount
+		reponse := d.processMount(r, false)
+		d.refCounts.MarkDirty()
+
+		// then unlock the state
+		return reponse
 	}
-	fstype = value
 
-	mountpoint, err := d.MountVolume(r.Name, fstype, "", isReadOnly, false)
-	if err != nil {
-		log.WithFields(
-			log.Fields{"name": r.Name, "error": err.Error()},
-		).Error("Failed to mount ")
-
-		refcnt, _ := d.decrRefCount(r.Name)
-		if refcnt == 0 {
-			log.Infof("Detaching %s - it is not used anymore", r.Name)
-			d.ops.Detach(r.Name, nil) // try to detach before failing the request for volume
-		}
-		return volume.Response{Err: err.Error()}
-	}
+	// if refcount has been succcessful,
+	// unlock and do usual normal mounting with refcounting as well
 
-	return volume.Response{Mountpoint: mountpoint}
+	return d.processMount(r, true)
 }
 
 // Unmount request from Docker. If mount refcount is drop to 0.
 // Unmount and detach from VM
 func (d *VolumeDriver) Unmount(r volume.UnmountRequest) volume.Response {
 	log.WithFields(log.Fields{"name": r.Name}).Info("Unmounting Volume ")
 
+	// lock the state
+	d.refCounts.StateMtx.Lock()
+	defer d.refCounts.StateMtx.Unlock()
+
+	if d.refCounts.GetSuccess() != true {
+		// if refcounting hasn't been succesful,
+		// no refcounting, no unmount. All unmounts are delayed
+		// until we succesfully populate the refcount map
+		d.refCounts.MarkDirty()
+		return volume.Response{Err: ""}
```

Comment:
Empty string means error is None. That's how we have indicated error to elsewhere to docker.